enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+---
+title: control-plane-machine-set-controller
+authors:
+  - "@JoelSpeed"
+reviewers:
+  - TBD
+approvers:
+  - "@sttts"
+  - "@soltysh"
+  - "@tkashem"
+  - "@hasbro17"
+api-approvers:
+  - TBD
+creation-date: 2022-01-11
+last-updated: 2022-02-02
+tracking-link:
+  - TBD
+replaces:
+  - "[/enhancements/machine-api/control-plane-machinesets.md](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/control-plane-machinesets.md)"
+  - "https://github.com/openshift/enhancements/pull/278"
+  - "https://github.com/openshift/enhancements/pull/292"
+---
+
+# Control Plane Machine Set Controller
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+To enable automated management of vertical scaling and replacement of Control Plane Machines, this proposal introduces
+a new resource and controller that will manage Control Plane Machine instances.
+
+## Motivation
+
+As OpenShift adoption increases and our managed services offerings become more popular, the manual effort required to
+scale Control Plane Machines as clusters grow or shrink is becoming a significant drain on the SRE Platform team
+managing OpenShift Dedicated (OSD), Red Hat OpenShift on AWS (ROSA), and Azure Red Hat OpenShift (ARO).
+
+The procedure to resize a Control Plane today is lengthy and very involved. It takes a significant amount of time for
+an OpenShift expert to perform. We also document this process for our end users, however due to the complexity of the
+procedure, there is discomfort in recommending this procedure to end users who may not be as familiar with the product.
+
+To ensure the long term usability of OpenShift, we must provide an automated way for users to scale their Control Plane
+as and when their cluster needs additional capacity.
+
+As HyperShift adoptions grows, HyperShift will solve the same issue by running hosted Control Planes within management
+clusters. However, HyperShift is not a suitable product for all OpenShift requirements (for example the first management
+cluster) and as such, we see that HA clusters will continue to be used and we must solve this problem to allow the
+continued adoption of HA clusters.
+
+### Goals
+
+* Provide a "best practices" approach to declarative management of Control Plane Machines
+* Allow users to "1-click" scale their Control Plane to large or smaller instances
```

Comment:
For compute nodes we don't currently support this. There's a section at the bottom that talks about the idea of a `StatefulMachineSet` where we discuss this a bit more in detail. The TL;DR is that we don't think a `StatefulMachineSet` is warranted, but, if we want the idea of 1 click changes, we should look into the `MachineDeployment` concept from Cluster API. Originally it was decided we didn't want this within OpenShift but perhaps we want to reconsider that decision.

It was our intention to actually borrow a lot of the `MachineDeployment` implementation/use it as inspiration for the logic within this new controller

Reply:
I guess the naming is what the issue is here, if it were for example, a ControlPlaneMachineDeployment, this would make the distinction more obvious as hopefully that would hint at its parallel being a MachineDeployment which we don't (yet) have.

I think we will just have to make sure in our documentation efforts that we are clear on what is the expectation here